velero-csi-plugin/csi_restorer.go

```diff
@@ -63,16 +66,35 @@ func (p *CSIRestorer) Execute(input *velero.RestoreItemActionExecuteInput) (*vel
 		return nil, errors.Errorf("Could not find volume snapshot name on PVC")
 	}
 
-	g := "snapshot.storage.k8s.io"
-	pvc.Spec.DataSource = &corev1api.TypedLocalObjectReference{
-		// This needs to be a pointer, since nil is used for the default group
-		APIGroup: &g,
-		Kind:     "VolumeSnapshot",
-		Name:     volumeSnapshotName,
+	for k := range annotations {
+		switch k {
+		case "velero.io/backup-name", "velero.io/volume-snapshot-name":
+			// keep these annotations
+		default:
+			// remove annotations like  pv.kubernetes.io/...
+			delete(annotations, k)
+		}
+	}
+
+	pvc.SetAnnotations(annotations)
+	sc := *pvc.Spec.StorageClassName
+	accessModes := pvc.Spec.AccessModes
+	resources := pvc.Spec.Resources
+
+	pvc.Spec = corev1api.PersistentVolumeClaimSpec{
+		StorageClassName: &sc,
+		DataSource: &core_v1.TypedLocalObjectReference{
+			APIGroup: &snapshotv1beta1api.SchemeGroupVersion.Group,
+			Kind:     "VolumeSnapshot",
+			Name:     volumeSnapshotName,
+		},
+		AccessModes: accessModes,
+		Resources:   resources,
 	}
 
 	pvcMap, err := runtime.DefaultUnstructuredConverter.ToUnstructured(&pvc)
 	if err != nil {
+		p.log.Errorf("failed to convert pvc into a map, %v", errors.WithStack(err))
```

Comment:
we should just return the error, not log *and* return. You can use `errors.Wrapf(...)` if you want to add an additional message to the error.

Reply:
leftovers from debugging.